core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -100,7 +100,7 @@
     // we use the hashed variant since we iterate over it and check removal and additions on existing keys
     private final LiveVersionMap versionMap;
 
-    private final Object[] dirtyLocks;
+    private final KeyedLock<BytesRef> keyedLock = new KeyedLock<>();
```

Comment:
Maybe should be called something with `dirty` in the name because all the methods refer to that name.